numba_mpi/api/requests.py

```diff
@@ -0,0 +1,199 @@
+    Status is currently not handled. Requires 'requests' parameter to be an
+    array of c-style pointers to MPI_Requests (e.g. created by
+    'create_requests_array' and popuated by 'isend'/'irecv').
+    """
+
+    status_buffer = create_status_buffer(requests.size)
+    status = _MPI_Waitall(
+        requests.size, requests.ctypes.data, status_buffer.ctypes.data
+    )
+
+    return status
+
+
+_MPI_Waitany = libmpi.MPI_Waitany
+_MPI_Waitany.restype = ctypes.c_int
+_MPI_Waitany.argtypes = [ctypes.c_int, _MpiRequestPtr, ctypes.c_void_p, _MpiStatusPtr]
+
+
+@numba.njit
+def waitany(requests):
+    """Wrapper for MPI_Waitany. Returns integer which is non-negative when call
+    succeeded and is the index of request that was completed, otherwise -
+    if result is negative - error occurred and its code may be obtained by
+    negating returned integer. Status is currently not handled.
+    Requires 'requests' parameter to be an array of c-style pointers to
+    MPI_Requests (e.g. created by 'create_requests_array' and popuated by
+    'isend'/'irecv').
+    """
+
+    status_buffer = create_status_buffer()
+    index = np.empty(1, dtype=np.intc)
+    status = _MPI_Waitany(
+        requests.size,
+        requests.ctypes.data,
+        index.ctypes.data,
+        status_buffer.ctypes.data,
+    )
+
+    if status > 0:
+        return -status
+    return index[0]
+
+
+# - Test* functions
+
+_MPI_Test = libmpi.MPI_Test
+_MPI_Test.restype = ctypes.c_int
+_MPI_Test.argtypes = [_MpiRequestPtr, ctypes.c_void_p, _MpiStatusPtr]
+
+
+@numba.njit
+def test(request):
+    """Wrapper for MPI_Test. Returns boolean flag indicating whether given
+    request is completed. Status is currently not handled. Requires
+    'request' parameter to be a c-style pointer to MPI_Request
+    (such as returned by 'isend'/'irecv').
+    """
+
+    status_buffer = create_status_buffer()
+    flag = np.empty(1, dtype=np.intc)
+    status = _MPI_Test(request.ctypes.data, flag.ctypes.data, status_buffer.ctypes.data)
+
+    assert status == 0
+
+    return flag[0] != 0
+
+
+_MPI_Testall = libmpi.MPI_Testall
+_MPI_Testall.restype = ctypes.c_int
+_MPI_Testall.argtypes = [ctypes.c_int, _MpiRequestPtr, ctypes.c_void_p, _MpiStatusPtr]
+
+
+@numba.njit
+def testall(requests):
+    """Wrapper for MPI_Testall. Returns boolean flag indicating whether all
+    requests in question are completed. Status is currently not handled.
+    Requires 'requests' parameter to be an array of c-style pointers to
+    MPI_Requests (e.g. created by 'create_requests_array' and popuated by
+    'isend'/'irecv').
+    """
+
+    status_buffer = create_status_buffer(requests.size)
+    flag = np.empty(1, dtype=np.intc)
+    status = _MPI_Testall(
+        requests.size, requests.ctypes.data, flag.ctypes.data, status_buffer.ctypes.data
+    )
+
+    assert status == 0
+
+    return flag[0] != 0
+
+
+@numba.experimental.jitclass([("value", numba.int32)])
```

Comment:
Numba jit classes have a very poor performance record. Here, it seems we could perhaps use a named tuple with `flag` and `index` fields?

Reply:
- removed jit class, changed result to tuple
- deferring to another PR that will address exception handling in the whole numba-mpi codebase